In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('Additional_data/network_teams.csv')
data.head()

Driver         From           To
0  adolf brudes      veritas      veritas
1   adolfo cruz       cooper       cooper
2  adrian sutil       spyker       spyker
3  adrian sutil       spyker  force india
4  adrian sutil  force india       sauber

In [3]:
transition_counts = data.groupby(['From', 'To']).size().reset_index(name='count')
total_transitions = transition_counts.groupby('From')['count'].sum().reset_index()
total_transitions.rename(columns={'count': 'total_out'}, inplace=True)
transition_matrix = transition_counts.merge(total_transitions, on='From')

transition_matrix['probability'] = transition_matrix['count'] / transition_matrix['total_out']

In [4]:
teams = sorted(set(data['From']).union(set(data['To'])))
team_to_index = {team: i for i, team in enumerate(teams)}
n = len(teams)
A = np.zeros((n, n))
for _, row in transition_matrix.iterrows():
    i, j = team_to_index[row['From']], team_to_index[row['To']]
    A[i, j] = row['probability']

A_df = pd.DataFrame(A, index=teams, columns=teams)
A_df.to_csv("Data_streamlit/q16.csv",index=False)
print("Final Stochastic Transition Matrix:")
print(A_df)


Final Stochastic Transition Matrix:
            adams       afm       ags  alfa romeo  alphatauri  alpine f1 team  \
adams         1.0  0.000000  0.000000    0.000000    0.000000        0.000000   
afm           0.0  0.666667  0.000000    0.000000    0.000000        0.000000   
ags           0.0  0.000000  0.333333    0.000000    0.000000        0.000000   
alfa romeo    0.0  0.000000  0.000000    0.346154    0.000000        0.000000   
alphatauri    0.0  0.000000  0.000000    0.000000    0.333333        0.166667   
...           ...       ...       ...         ...         ...             ...   
watson        0.0  0.000000  0.000000    0.000000    0.000000        0.000000   
wetteroth     0.0  0.000000  0.000000    0.000000    0.000000        0.000000   
williams      0.0  0.000000  0.000000    0.015385    0.015385        0.000000   
wolf          0.0  0.000000  0.000000    0.000000    0.000000        0.000000   
zakspeed      0.0  0.000000  0.000000    0.000000    0.000000        0.00

Driver transitions

In [5]:
driver = 'max verstappen'

In [8]:
results = pd.read_csv('final_dataset/results.csv')
driver = pd.read_csv('final_dataset/drivers.csv')
constructor = pd.read_csv('final_dataset/constructors.csv')
races = pd.read_csv('final_dataset/races.csv')

results = results.merge(races[['raceId', 'year']], on='raceId', how='left')
results = results.merge(driver[['driverId','forename', 'surname']], on='driverId', how='left')
results = results.merge(constructor[['constructorId', 'name']], on='constructorId', how='left')
results['forename'] = results['forename'].astype('str')
results['surname'] = results['surname'].astype('str')
results['Name']  = results['forename']+" "+results['surname']
results.drop(columns=['forename','surname'],inplace=True)
results.rename(columns={'name':'Constructor_name'},inplace=True)

In [10]:
team = 'red bull'

Finding probability of transitioning from red bull to other teams and finding max probability from that

In [12]:
transitions = A_df.loc[team].sort_values(ascending=False)
top_teams = transitions.head(3)
top_teams.to_csv('Data_streamlit/q16_2.csv',index=False)
top_teams


toro rosso    0.333333
ferrari       0.111111
williams      0.111111
Name: red bull, dtype: float64

From the probability estimate it is seen that there is a probability of 0.33 for the driver to move from red bull to toro rosso